servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/ConcurrentStack.java

```diff
@@ -0,0 +1,128 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.concurrent.api;
+
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import java.util.function.Consumer;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.concurrent.internal.ThrowableUtils.catchUnexpected;
+import static io.servicetalk.utils.internal.PlatformDependent.throwException;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.atomic.AtomicReferenceFieldUpdater.newUpdater;
+
+final class ConcurrentStack<T> {
+    private static final Node<?> CLOSED = new Node<>();
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<ConcurrentStack, Node> topUpdater =
+            newUpdater(ConcurrentStack.class, Node.class, "top");
+    @Nullable
+    private volatile Node<T> top;
+
+    boolean push(T item) {
+        final Node<T> newTop = new Node<>(item);
+        Node<T> oldTop;
+        do {
+            oldTop = top;
+            if (oldTop == CLOSED) {
+                return false;
+            }
+            newTop.next = oldTop;
+        } while (!topUpdater.compareAndSet(this, oldTop, newTop));
+        return true;
+    }
+
+    @Nullable
+    T pop() {
+        for (;;) {
+            final Node<T> oldTop = top;
+            if (oldTop == null) {
+                return null;
+            } else if (topUpdater.compareAndSet(this, oldTop, oldTop.next)) {
+                final T item = oldTop.item;
+                if (item != null) {
+                    return item;
+                }
+            }
+        }
+    }
+
+    /**
+     * Best effort removal of {@code item} from this stack.
+     * @param item The item to remove.
+     * @return {@code true} if the item was found in this stack and marked for removal. The "relaxed" nature of
+     * this method means {@code true} might be returned in the following scenarios without external synchronization:
+     * <ul>
+     *     <li>invoked multiple times with the same {@code item} from different threads</li>
+     *     <li>{@link #pop()} removes this item from another thread</li>
+     * </ul>
+     */
+    boolean relaxedRemove(T item) {
+        requireNonNull(item);
+        Node<T> currTop = top;
+        while (currTop != null) {
+            if (item.equals(currTop.item)) {
+                currTop.item = null; // best effort null out the item. pop/close will discard the Node later.
+                return true;
+            } else {
+                currTop = currTop.next;
+            }
+        }
+        return false;
+    }
+
+    void close(Consumer<T> closer) {
+        @SuppressWarnings("unchecked")
+        Node<T> oldTop = topUpdater.getAndSet(this, closedNode());
+        Throwable delayedCause = null;
+        while (oldTop != null && oldTop != CLOSED) {
```

Comment:
`CLOSED` can be only at the top. Consider removing this check from the loop and fast exit if `oldTop == CLOSED`.